api/backend/blueflood/blueflood.go

```diff
@@ -198,8 +198,56 @@ func bluefloodResolution(r int64) string {
 	return Resolution1440Min
 }
 
-func resample(points []float64, currentResolution int64, expectedTimerange api.Timerange, sampleMethod api.SampleMethod) ([]float64, error) {
-	return nil, errors.New("Not implemented")
+// Resamples the original data returned from Blueflood. Returns raw values
+// corresponding to the `timerange`.
+//
+// Snaps data left to the `timerange` buckets. Resampling buckets using data
+// to the right of the timestamp (i.e. the data point at time 123000 with
+// resolution 60 is generated from [123000, 123060). If there is no data, that
+// point is filled with NaN (note: this means upsampling doesn't work).
+//
+// Assumes `points` is sorted chronologically.
+func resample(points []MetricPoint, timerange api.Timerange, sampleMethod api.SampleMethod) ([]float64, error) {
+	series := make([]float64, timerange.Slots())
+
+	// Cursor into `series`
+	idx := 0
+	// Cursors into `points` for working data windows
+	windowStart, windowEnd := 0, 0
+	// Generate points at timestamps corresponding to the timerange
+	for t := timerange.Start(); t <= timerange.End(); t, idx = t+timerange.Resolution(), idx+1 {
+		// Set up our series window
+		for ; points[windowStart].Timestamp < t; windowStart++ {
+		}
+		// No data points in [t, t + resolution] becomes NaN
+		if points[windowStart].Timestamp >= (t + timerange.Resolution()) {
+			series[idx] = math.NaN()
+			continue
+		}
+		for windowEnd = windowStart; windowEnd < len(points)-1 && points[windowEnd+1].Timestamp < (t+timerange.Resolution()); windowEnd++ {
+		}
+
+		switch sampleMethod {
+		case api.SampleMean:
+			// Uses weighted average over [t, t + resolution)
+			totalPoints, runningTotal := 0, 0.0
+			for i := windowStart; i <= windowEnd; i++ {
+				totalPoints += points[i].Points
+				runningTotal += points[i].Average * float64(points[i].Points)
```

Comment:
i dislike the implicit assumption to `Average` is weird. maybe use nathan's solution to extract fields?